File: smash-ranks-android/app/src/main/java/com/garpr/android/models/FullPlayer.java
package com.garpr.android.models;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

public class FullPlayer extends AbsPlayer implements Parcelable {

    @Nullable
    @SerializedName("aliases")
    private ArrayList<String> mAliases;

    @Nullable
    @SerializedName("regions")
    private ArrayList<String> mRegions;

    @Nullable
    @SerializedName("ratings")
    private Ratings mRatings;


    @Nullable
    public ArrayList<String> getAliases() {
        return mAliases;
    }

    @Override
    public Kind getKind() {
        return Kind.FULL;
    }

    @Nullable
    public Ratings getRatings() {
        return mRatings;
    }

    @Nullable
    public ArrayList<String> getRegions() {
        return mRegions;
    }

    public boolean hasAliases() {
        return mAliases != null && !mAliases.isEmpty();
    }

    public boolean hasRatings() {
        return mRatings != null && !mRatings.isEmpty();
    }

    public boolean hasRegions() {
        return mRegions != null && !mRegions.isEmpty();
    }

    @Override
    protected void readFromParcel(final Parcel source) {
        super.readFromParcel(source);
        mAliases = source.createStringArrayList();
        mRegions = source.createStringArrayList();
        mRatings = source.readParcelable(Ratings.class.getClassLoader());
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        super.writeToParcel(dest, flags);
        dest.writeStringList(mAliases);
        dest.writeStringList(mRegions);
        dest.writeParcelable(mRatings, flags);
    }

    public static final Creator<FullPlayer> CREATOR = new Creator<FullPlayer>() {
        @Override
        public FullPlayer createFromParcel(final Parcel source) {
            final FullPlayer p = new FullPlayer();
            p.readFromParcel(source);
            return p;
        }

        @Override
        public FullPlayer[] newArray(final int size) {
            return new FullPlayer[size];
        }
    };

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/models/FullPlayer.kt
package com.garpr.android.models

import android.os.Parcel
import android.os.Parcelable
import com.garpr.android.extensions.createParcel
import com.garpr.android.extensions.readRatingsMap
import com.garpr.android.extensions.requireString
import com.garpr.android.extensions.writeRatingsMap
import com.google.gson.annotations.SerializedName

class FullPlayer(
        id: String,
        name: String,
        @SerializedName("aliases") val aliases: List<String>? = null,
        @SerializedName("regions") val regions: List<String>? = null,
        @SerializedName("ratings") val ratings: Map<String, Rating>? = null
) : AbsPlayer(
        id,
        name
), Parcelable {

    companion object {
        @JvmField
        val CREATOR = createParcel {
            FullPlayer(
                    it.requireString(),
                    it.requireString(),
                    it.createStringArrayList(),
                    it.createStringArrayList(),
                    it.readRatingsMap()
            )
        }
    }

    override val kind
        get() = Kind.FULL

    val uniqueAliases: Array<String>?
        get() {
            if (aliases == null || aliases.isEmpty()) {
                return null
            }

            val uniqueAliases = mutableListOf<String>()

            for (alias in aliases) {
                if (alias.isNotBlank() && !alias.equals(name, true)) {
                    var add = true

                    for (uniqueAlias in uniqueAliases) {
                        if (alias.equals(uniqueAlias, true)) {
                            add = false
                            break
                        }
                    }

                    if (add) {
                        uniqueAliases.add(alias)
                    }
                }
            }

            return if (uniqueAliases.isEmpty()) {
                null
            } else {
                uniqueAliases.toTypedArray()
            }
        }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeStringList(aliases)
        dest.writeStringList(regions)
        dest.writeRatingsMap(ratings)
    }

}
